src/hopen/syntax/mustache.cljc

```diff
@@ -0,0 +1,434 @@
+     ;; it makes use of some global state, which may mess up the result.
+     ;;
+     ;; This is an altered version of cljs' re-seq.
+     (letfn [(re-seq-pos* [re s current-offset]
+               (when-some [matches (.exec re s)]
+                 (let [match-str (aget matches 0)
+                       start (+ current-offset (.-index matches))
+                       end (+ start (count match-str))]
+                   (cons {:start start
+                          :end end
+                          :groups (if (== (.-length matches) 1)
+                                    match-str
+                                    (vec matches))}
+                         (lazy-seq
+                          (let [post-idx (+ (.-index matches)
+                                            (max 1 (.-length match-str)))]
+                            (when (<= post-idx (.-length s))
+                              (re-seq-pos* re (subs s post-idx) (+ current-offset post-idx)))))))))]
+     (if (string? string)
+         (re-seq-pos* pattern string 0)
+         (throw (js/TypeError. "re-seq-pos must match against a string."))))))
+
+(defn conj-last
+  "Conj `x` into the last element of the `coll`.
+  `coll` should be a seq of seqs."
+  [coll x]
+  (conj (pop coll) (conj (peek coll) x)))
+
+(defn conj-some
+  [coll & xs]
+  (->> xs
+       (filter some?)
+       (apply conj coll)))
+
+(defn deep-merge
+  [a b]
+  (if (map? a)
+    (into a (for [[k v] b]
+              [k (deep-merge (a k) v)]))
+    b))
+
+;;------------------------------------------------------------------------------
+;; Namespace specific functions
+;;------------------------------------------------------------------------------
+
+(defn- compile-regex [start-delim end-delim]
+  (if (and (= start-delim "{{")
+           (= end-delim "}}"))
+    #"\{\{\{(.*?)\}\}\}|\{\{(.*?)\}\}"
+    (re-pattern
+     (str
+      (util/re-quote start-delim)
+      "\\{(.*?)\\}"
+      (util/re-quote start-delim)
+      "|"
+      (util/re-quote start-delim)
+      "(.*?)"
+      (util/re-quote end-delim)))))
+
+(defn- parse-delim-change-tag
+  "Try to parse the tag-text as a delimiter change command.
+  Returns '(start-delim end-delim) or nil."
+  [tag-text]
+  (let [result (re-matches #"=(.*)\s+(.*)=" tag-text)]
+    (when (= 3 (count result))
+      (drop 1 result))))
+
+(defn- tag--text [tag]
+  (get-in tag [:groups 1]))
+
+(defn- retrieve-all-tags
+  "Retrieve a list of all tags and their offsets.
+  Any delimiter changes will be dealt with here.
+
+  A tag is represented by a single entry returned by `re-seq-pos`,
+  which looks like:
+  {:start 'character offset into text where the tag starts'
+   :end 'character offset into text where the tag ends'
+
+   :groups ['full text of tag with delimiters' 'text of tag without delimiters']}"
+  [text]
+
+  (let [update-tags-with-offset (fn [offset tags]
+                                  (if (zero? offset)
```

Comment:
IHMO, premature optimization, the test to zero can be removed.

Reply:
The rest of the code updates the tags' `:start` and `:end` fields with the `offset`. Why should we map over the collection and add 0 to every field when we don't need to?

Reply:
There is a compromise to do between performance and source code readability.
If you think that it matters to avoid adding zero, keep it as it is.

Reply:
okay